confluentinc/vscode
# cleanup/fix: track `Disposable`s better during extension activation and prevent reactivation/upgrade errors

## Summary of Changes



Closes #210, #476, and #520.

Two main problems before this PR:
1. we weren't adding a lot of our `Disposable`s to the extension context `subscriptions` to be (automatically) cleaned up during extension deactivation, so we'd get errors about views already being registered -- in this example, the Resources view's refresh command was tracked: https://github.com/confluentinc/vscode/blob/33b7a8e63f3b37427d64f04edbb5bd84e4f34b73/src/extension.ts#L282-L288
    ...but not the actual ResourceViewProvider's `treeView`: https://github.com/confluentinc/vscode/blob/33b7a8e63f3b37427d64f04edbb5bd84e4f34b73/src/viewProviders/resources.ts#L71
2. our static `Confluent Cloud` and `Local` container items in the Resources view were using static `id` values, which apparently can't be used repeatedly during extension activation
    - this isn't an issue for the majority of our other items because we aren't explicitly setting `id` (but we may need it for [here](https://github.com/confluentinc/vscode/blob/33b7a8e63f3b37427d64f04edbb5bd84e4f34b73/src/models/schema.ts#L75) @jlrobins)

To address the first problem, we're now returning more `Disposable`s during extension activation and adding them to the extension context subscriptions within [_activateExtension()](https://github.com/confluentinc/vscode/blob/djs/registration-cleanup/src/extension.ts#L100). In order to get here, many of our classes needed to start keeping track of their own disposables, which includes
- [ResourceViewProvider](https://github.com/confluentinc/vscode/blob/djs/registration-cleanup/src/viewProviders/resources.ts#L80)
- [TopicViewProvider](https://github.com/confluentinc/vscode/blob/djs/registration-cleanup/src/viewProviders/topics.ts#L70)
- [SchemasViewProvider](https://github.com/confluentinc/vscode/blob/djs/registration-cleanup/src/viewProviders/schemas.ts#L63)
- [ResourceLoader](https://github.com/confluentinc/vscode/blob/djs/registration-cleanup/src/storage/resourceLoader.ts#L73)
- ...and a few other instances of single-use event listeners

## How do you even test that this works?

- Pull this branch locally
- Run `gulp bundle` to get a .vsix file in the `out/` directory, then save that off somewhere else as your "current version" extension
- Run `gulp bundle` to get a(n even newer) .vsix file in the `out/` directory
- Install one of those extensions, activate it by normal extension use
- Without uninstalling the current extension, install whatever .vsix file you didn't install

Should show a clean upgrade path that looks something like this:


https://github.com/user-attachments/assets/9792f6cd-cfbc-4561-8b13-619fe735613a




## Pull request checklist

Please check if your PR fulfills the following (if applicable):

##### Tests


##### Other


  ```shell
  gulp clicktest
  ```
